bigquery_etl/view/generate_stable_views.py

```diff
@@ -0,0 +1,251 @@
+    "/archive/generated-schemas.tar.gz"
+)
+
+VIEW_QUERY_TEMPLATE = """\
+-- Generated by bigquery_etl.view.generate_stable_views
+
+SELECT
+  * REPLACE(
+    {replacements})
+FROM
+  `{target}`
+"""
+
+VIEW_METADATA_TEMPLATE = """\
+# Generated by bigquery_etl.view.generate_stable_views
+---
+friendly_name: Historical Pings for ``{document_namespace}/{document_type}`
+description: |-
+  A historical view of pings sent for the
+  `{document_namespace}/{document_type}`
+  document type.
+
+  This view is guaranteed to contain only complete days
+  (per `submission_timestamp`)
+  and to contain only one row per distinct `document_id` within a given date.
+
+  Clustering fields: `normalized_channel`, `sample_id`
+"""
+
+parser = ArgumentParser(description=__doc__)
+parser.add_argument(
+    "--sql-dir", default="sql/", help="The path where generated SQL files are stored."
+)
+parser.add_argument(
+    "--target-project",
+    default="moz-fx-data-shared-prod",
+    help="The project where the stable tables live.",
+)
+standard_args.add_log_level(parser)
+standard_args.add_parallelism(parser)
+
+
+@dataclass
+class SchemaFile:
+    """Container for metadata about a JSON schema and corresponding BQ table."""
+
+    schema_id: str
+    bq_dataset_family: str
+    bq_table: str
+    document_namespace: str
+    document_type: str
+    document_version: int
+
+    @property
+    def bq_table_unversioned(self):
+        """Return table_id with version suffix stripped."""
+        return "_".join(self.bq_table.split("_")[:-1])
+
+    @property
+    def stable_table(self):
+        """Return BQ stable table name in <dataset>.<table> form."""
+        return f"{self.bq_dataset_family}_stable.{self.bq_table}"
+
+    @property
+    def user_facing_view(self):
+        """Return user-facing view name in <dataset>.<view> form."""
+        return f"{self.bq_dataset_family}.{self.bq_table_unversioned}"
+
+    @property
+    def sortkey(self):
+        """Return variant of stable_table with zero-padded version for sorting."""
+        return (
+            "_".join(self.stable_table.split("_")[:-1]) + f"{self.document_version:04d}"
+        )
+
+
+def write_view_if_not_exists(target_project: str, sql_dir: Path, schema: SchemaFile):
+    """If a view.sql does not already exist, write one to the target directory."""
+    target_dir = (
+        sql_dir
+        / target_project
+        / schema.bq_dataset_family
+        / schema.bq_table_unversioned
+    )
+    target_file = target_dir / "view.sql"
+
+    if target_file.exists():
+        return
+
+    # Exclude doctypes maintained in separate projects.
+    if schema.bq_dataset_family.startswith("pioneer"):
+        return
+    if schema.bq_dataset_family.startswith("rally"):
+        return
+
+    full_source_id = f"{target_project}.{schema.stable_table}"
+    full_view_id = f"{target_project}.{schema.user_facing_view}"
+    replacements = ["mozfun.norm.metadata(metadata)" " AS metadata"]
```

Comment:
```suggestion
    replacements = ["mozfun.norm.metadata(metadata) AS metadata"]
```
Is there a reason for splitting these replacement strings?